File: get_refresh_token.py
#!/usr/bin/env python3
"""
Spotify Refresh Token Generator for ESP32
Run this script on your computer to log in to Spotify and retrieve your SPOTIFY_REFRESH_TOKEN.
Zero external dependencies required (uses built-in Python standard libraries).
"""

import sys
import urllib.parse
import urllib.request
import json
import base64
import webbrowser
from http.server import HTTPServer, BaseHTTPRequestHandler

REDIRECT_URI = "http://localhost:8888/callback"
SCOPES = "user-read-currently-playing user-read-playback-state"

auth_code = None

class CallbackHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        global auth_code
        parsed_path = urllib.parse.urlparse(self.path)
        query = urllib.parse.parse_qs(parsed_path.query)

        if "code" in query:
            auth_code = query["code"][0]
            self.send_response(200)
            self.send_header("Content-Type", "text/html")
            self.end_headers()
            self.wfile.write(b"<h1>Authorization Successful!</h1><p>You can close this tab and return to your terminal.</p>")
        else:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"<h1>Authorization Failed!</h1>")

    def log_message(self, format, *args):
        return  # Suppress default server logs

def main():
    print("=" * 65)
    print(" Spotify OAuth 2.0 Refresh Token Generator for ESP32")
    print("=" * 65)

    client_id = input("\nEnter your Spotify Client ID: ").strip()
    client_secret = input("Enter your Spotify Client Secret: ").strip()

    if not client_id or not client_secret:
        print("[Error] Client ID and Client Secret cannot be empty!")
        sys.exit(1)

    # Construct Auth URL
    auth_params = {
        "client_id": client_id,
        "response_type": "code",
        "redirect_uri": REDIRECT_URI,
        "scope": SCOPES,
    }
    auth_url = "https://accounts.spotify.com/authorize?" + urllib.parse.urlencode(auth_params)

    print("\n[Step 1] Opening browser to authenticate with Spotify...")
    print(f"If browser doesn't open automatically, visit:\n{auth_url}\n")
    webbrowser.open(auth_url)

    # Start local HTTP server to receive callback code
    server_address = ("", 8888)
    httpd = HTTPServer(server_address, CallbackHandler)
    print("[Step 2] Waiting for authorization callback on http://localhost:8888/callback ...")
    
    while auth_code is None:
        httpd.handle_request()

    print("[Success] Received Authorization Code!")

    # Exchange Authorization Code for Refresh Token
    print("\n[Step 3] Exchanging Code for Refresh Token...")
    token_url = "https://accounts.spotify.com/api/token"
    
    auth_header = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()
    headers = {
        "Authorization": f"Basic {auth_header}",
        "Content-Type": "application/x-www-form-urlencoded"
    }

    data = urllib.parse.urlencode({
        "grant_type": "authorization_code",
        "code": auth_code,
        "redirect_uri": REDIRECT_URI
    }).encode("utf-8")

    req = urllib.request.Request(token_url, data=data, headers=headers, method="POST")

    try:
        with urllib.request.urlopen(req) as response:
            res_body = response.read().decode("utf-8")
            token_json = json.loads(res_body)

            refresh_token = token_json.get("refresh_token")
            access_token = token_json.get("access_token")

            print("\n" + "=" * 65)
            print(" YOUR SPOTIFY REFRESH TOKEN:")
            print("=" * 65)
            print(f"\n{refresh_token}\n")
            print("=" * 65)
            print("Copy the token above and paste it into config.h:")
            print('#define SPOTIFY_REFRESH_TOKEN "' + str(refresh_token) + '"')
            print("=" * 65 + "\n")

    except Exception as e:
        print(f"[Error] Failed to obtain refresh token: {e}")

if __name__ == "__main__":
    main()

File: get_refresh_token.js
#!/usr/bin/env node
/**
 * Spotify Refresh Token Generator for ESP32 (Node.js version)
 * Zero external dependencies (uses Node.js built-in modules).
 */

const http = require('http');
const https = require('https');
const url = require('url');
const readline = require('readline');
const { exec } = require('child_process');

const REDIRECT_URI = 'http://localhost:8888/callback';
const SCOPES = 'user-read-currently-playing user-read-playback-state';

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function question(query) {
    return new Promise(resolve => rl.question(query, resolve));
}

function openBrowser(targetUrl) {
    const start = process.platform === 'darwin' ? 'open' :
                  process.platform === 'win32' ? 'start ""' : 'xdg-open';
    exec(`${start} "${targetUrl}"`);
}

async function main() {
    console.log('='.repeat(65));
    console.log(' Spotify OAuth 2.0 Refresh Token Generator for ESP32');
    console.log('='.repeat(65));

    const clientId = (await question('\nEnter your Spotify Client ID: ')).trim();
    const clientSecret = (await question('Enter your Spotify Client Secret: ')).trim();

    if (!clientId || !clientSecret) {
        console.error('[Error] Client ID and Client Secret cannot be empty!');
        rl.close();
        process.exit(1);
    }

    const authUrl = `https://accounts.spotify.com/authorize?${new URLSearchParams({
        client_id: clientId,
        response_type: 'code',
        redirect_uri: REDIRECT_URI,
        scope: SCOPES
    }).toString()}`;

    console.log('\n[Step 1] Opening browser to authenticate with Spotify...');
    console.log(`If browser does not open automatically, visit:\n${authUrl}\n`);
    openBrowser(authUrl);

    console.log('[Step 2] Waiting for authorization callback on http://localhost:8888/callback ...');

    const server = http.createServer(async (req, res) => {
        const reqUrl = url.parse(req.url, true);
        if (reqUrl.pathname === '/callback') {
            const code = reqUrl.query.code;
            if (code) {
                res.writeHead(200, { 'Content-Type': 'text/html' });
                res.end('<h1>Authorization Successful!</h1><p>You can close this tab and return to your terminal.</p>');
                server.close();

                console.log('[Success] Received Authorization Code!');
                console.log('\n[Step 3] Exchanging Code for Refresh Token...');

                // Request token
                const postData = new URLSearchParams({
                    grant_type: 'authorization_code',
                    code: code,
                    redirect_uri: REDIRECT_URI
                }).toString();

                const authHeader = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');

                const tokenReq = https.request('https://accounts.spotify.com/api/token', {
                    method: 'POST',
                    headers: {
                        'Authorization': `Basic ${authHeader}`,
                        'Content-Type': 'application/x-www-form-urlencoded',
                        'Content-Length': Buffer.byteLength(postData)
                    }
                }, (tokenRes) => {
                    let body = '';
                    tokenRes.on('data', chunk => body += chunk);
                    tokenRes.on('end', () => {
                        try {
                            const data = JSON.parse(body);
                            if (data.refresh_token) {
                                console.log('\n' + '='.repeat(65));
                                console.log(' YOUR SPOTIFY REFRESH TOKEN:');
                                console.log('='.repeat(65));
                                console.log(`\n${data.refresh_token}\n`);
                                console.log('='.repeat(65));
                                console.log('Copy the token above and paste it into config.h:');
                                console.log(`#define SPOTIFY_REFRESH_TOKEN "${data.refresh_token}"`);
                                console.log('='.repeat(65) + '\n');
                            } else {
                                console.error('[Error] Failed to get refresh token:', data);
                            }
                        } catch (e) {
                            console.error('[Error] Failed to parse response:', e.message, body);
                        }
                        rl.close();
                    });
                });

                tokenReq.on('error', (e) => {
                    console.error('[Error] Request failed:', e.message);
                    rl.close();
                });

                tokenReq.write(postData);
                tokenReq.end();
            } else {
                res.writeHead(400, { 'Content-Type': 'text/html' });
                res.end('<h1>Authorization Failed! No code received.</h1>');
            }
        }
    });

    server.listen(8888);
}

main();
